parlai/agents/bart/modules.py

```diff
@@ -56,7 +56,19 @@ def reorder_decoder_incremental_state(
         """
         Incremental state is weird to handle when we seed decoder with two inputs
         initially.
-
-        We leave as a future exercise.
         """
-        return None
+        # we only have this method called when it's actually being used
+        assert incremental_state is not None
+        assert len(incremental_state) > 0
+
+        for layer in incremental_state.keys():
+            incr_state_l = incremental_state[layer]
+            assert 'self_attn' in incr_state_l
+            assert 'prev_mask' in incr_state_l['self_attn']
+            self_attn_mask = incr_state_l['self_attn']['prev_mask']
+            # check this is on the very first run with incremental state
+            if self_attn_mask.ndim == 3 and tuple(self_attn_mask.shape[1:]) == (2, 2):
+                # cut off the inappropriate incremental state
+                incr_state_l['self_attn']['prev_mask'] = self_attn_mask[:, -1:, :]
```

Comment:
Hmm so I suppose the problem was that dim=1 of `prev_mask` included both start tokens, whereas `prev_key` and `prev_value` only had 1 each?

Reply:
Problem was that prev mask was a Bx2x2 tensor instead of a Bx2x1. The mask here appears the be accumulating what the new token is allowed to attend to (everything)

Reply:
Assuming you mean Bx2xD vs. Bx1xD in this explanation? (fits the line too)